lib/mllp/client.ex

```diff
@@ -0,0 +1,444 @@
+    options1 = Map.merge(state.send_opts, options)
+    telemetry(:sending, %{}, state)
+
+    payload = maybe_wrap_message(message, options1)
+
+    case state.tcp.send(state.socket, payload) do
+      :ok ->
+        case maybe_receive_reply(state, payload, options1) do
+          {:ok, :noop} ->
+            {:reply, {:ok, :sent}, state}
+
+          {_ok_or_error, _ack_type, %MLLP.Ack{}} = reply ->
+            {:reply, reply, state}
+
+          {:ok, _} = reply ->
+            {:reply, reply, state}
+
+          {:error, reason} ->
+            telemetry(
+              :status,
+              %{
+                status: :disconnected,
+                error: format_error(reason),
+                context: "receive ACK failure"
+              },
+              state
+            )
+
+            new_state = maintain_reconnect_timer(state)
+            reply = {:error, new_error(:recv, reason)}
+            {:reply, reply, new_state}
+        end
+
+      {:error, reason} ->
+        telemetry(
+          :status,
+          %{status: :disconnected, error: format_error(reason), context: "send message failure"},
+          state
+        )
+
+        new_state = maintain_reconnect_timer(state)
+        reply = {:error, new_error(:send, reason)}
+        {:reply, reply, new_state}
+    end
+  end
+
+  def handle_info(:timeout, state) do
+    new_state =
+      state
+      |> stop_connection(:timeout, "timeout message")
+      |> attempt_connection()
+
+    {:noreply, new_state}
+  end
+
+  def handle_info(unknown, state) do
+    Logger.warn("Unknown kernel message received => #{inspect(unknown)}")
+    {:noreply, state}
+  end
+
+  def terminate(reason, state) do
+    Logger.error("Client socket terminated. Reason: #{inspect(reason)} State #{inspect(state)}")
+    stop_connection(state, reason, "process terminated")
+  end
+
+  defp stop_connection(%State{} = state, error, context) do
+    if state.socket != nil do
+      telemetry(
+        :status,
+        %{status: :disconnected, error: format_error(error), context: context},
+        state
+      )
+
+      state.tcp.close(state.socket)
+    end
+
+    ensure_pending_reconnect_cancelled(state)
+  end
+
+  defp ensure_pending_reconnect_cancelled(%{pending_reconnect: nil} = state), do: state
+
+  defp ensure_pending_reconnect_cancelled(state) do
+    :ok = Process.cancel_timer(state.pending_reconnect, info: false)
+    %{state | pending_reconnect: nil}
+  end
+
+  defp attempt_connection(%State{} = state) do
+    telemetry(:status, %{status: :connecting}, state)
+    opts = [:binary, {:packet, 0}, {:active, false}] ++ state.socket_opts ++ state.tls_opts
+
+    case state.tcp.connect(state.address, state.port, opts, 2000) do
+      {:ok, socket} ->
+        state1 = ensure_pending_reconnect_cancelled(state)
+        telemetry(:status, %{status: :connected}, state1)
+        %{state1 | socket: socket, connect_failure: nil}
+
+      {:error, reason} ->
+        message = format_error(reason)
```

Comment:
We now log on a  failure to connect always, we also store the last connect failure error in state in the case the caller tries to send when we the connection is still trying to be established.